Factor shared coordinate map code and metafunctions

nilsdeppe opened the pull request:
## Proposed changes

Some of the metafunctions used by `CoordinateMap` is useful in time-dependent product maps, and possibly other time-dependent composition-type maps. The same is true for some of the helper functions for invoking different maps.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Factor metafunctions to check for map time dependence

nilsdeppe committed:
Factor coordinate map helpers

These and similar functions will also be used for time-dependent product maps.

moxcodes commented on src/Domain/CoordinateMaps/TimeDependentHelpers.hpp:
```diff
@@ -0,0 +1,43 @@
+CREATE_IS_CALLABLE(jacobian)
+}  // namespace detail
+
+/// Check if the call to the Jacobian of the coordinate map is time-dependent
```
Is this the sort of metafunction that will be needed for many of the CoordinateMap member functions? If so, it could be worth it to write another macro for creating the time dependency check.

nilsdeppe replied:
I'm a bit confused by your question. This check should work for all coordinate maps that we will ever write, no need for more type traits (maybe Hessian since we will probably need those). Does that clarify things?

moxcodes replied:
Let me see if I understand: This metafunction is keyed to examining the `jacobian` member function to determine if it is time dependent. Do I understand correctly that we'll only want to check the `jacobian` and not the `inverse` or `inv_jacobian`, but maybe a Hessian which could be addressed in later additions?

kidder replied:
@moxcodes if either the map or Jacobian are time-dependent, the corresponding inverses must also be so we need only one type trait for each...  and similarly for the Hessians when they are added

kidder replied:
might be worth adding the above to the doxygen for the type traits

moxcodes commented on src/Domain/CoordinateMaps/CoordinateMapHelpers.hpp:
```diff
@@ -0,0 +1,61 @@
+        functions_of_time,
+    const std::true_type
+    /*is_time_dependent*/) {
+  ASSERT(not functions_of_time.empty(),
```
:+1:

moxcodes commented on tests/Unit/Domain/CoordinateMaps/Test_TimeDependentHelpers.cpp:
```diff
@@ -0,0 +1,149 @@
+              "Failed testing is_map_time_dependent_t");
+static_assert(not is_map_time_dependent_t<TimeIndepMap<2>>::value,
+              "Failed testing is_map_time_dependent_t");
+static_assert(not is_map_time_dependent_t<TimeIndepMap<3>>::value,
```
feels very thorough, but I don't think extra static asserts ever hurt anyone :P

moxcodes reviewed (dismissed):
One question about generalizing the metafunction if we're going to need similar versions for other things. Otherwise, looks good to me.

moxcodes commented on src/Domain/CoordinateMaps/CoordinateMapHelpers.hpp:
```diff
@@ -0,0 +1,61 @@
+        return not isnan;
+      }(),
+      "The time must not be NaN for time-dependent maps.");
+  *t_map_point = the_map(*t_map_point, t, functions_of_time);
```
Sorry, got confused about this line while reviewing the other PR. This function seems to just forward, but this line makes it seem like there's little to be gained from its pass-by-pointer implementation because the `t_map_point` just gets passed by value (and presumably causes a new allocation) in the forward. Is that right?

nilsdeppe replied:
Yes. We will need to reevaluate the use of `not_null`s in time-dependent coordinate maps once we have moving meshes because currently there are a lot of internal allocations that don't matter in the time-independent case but will matter once the maps are time-dependent. Does that sound good?

moxcodes replied:
Ah, yeah, that sounds good; I'm happy to make the allocation optimization a future topic.

kidder reviewed (commented):
LGTM, other than the optional addition to the documentation.

kidder commented:
@gsb76 will you be able to look at this soon?

nilsdeppe committed:
Clarify dox comments for TimeDep coord map helpers

nilsdeppe commented:
Rebased on develop and expanded the documentation to make clear that the checks are for both the Jac/map and inverse calls. I added a new commit for the dox since I want to change as little as possible in refactor commits.
